test_player/views.py

```diff
@@ -0,0 +1,352 @@
+
+        # updating session with relevant data
+        session_data = self.request.session[current_quiz]
+        if form.has_changed():
+            session_data.update(
+                {question_number: answer}
+            )
+        self.request.session[current_quiz] = session_data
+
+        return super(TestPlayer, self).form_valid(form)
+
+    def _setting_initial_session_data(self):
+        """
+        Method, which set default initial values (used only once during the
+         first request).
+
+        :var questions(queryset object): which contain ordered questions
+         numbers;
+        :var initial_questions(dict): which contain initial key(questions),
+         values(answers) for questions.
+
+        """
+        current_quiz, _ = self.get_quiz_data()
+
+        questions = Questions.objects.filter(quiz=current_quiz).values_list(
+            'question_number', flat=True).order_by('question_number')
+        initial_questions = dict.fromkeys(questions)
+
+        self.request.session.setdefault(current_quiz, None)
+        self.request.session[current_quiz] = initial_questions
+
+    def _saving_user_answers(self, current_quiz):
+        """
+        Method, which save user answers to database such as json object.
+
+        :var session_data(dict): with session data from session object;
+        :var session_user(int): current logged in user who passes test.
+
+        """
+        session_user = self.request.session['_auth_user_id']
+        session_data = dict(self.request.session[current_quiz])
+        result = []
+
+        for answer in session_data:
+            result.append(int(session_data.get(answer)))
+        result = json.dumps(result, ensure_ascii=False)
+
+        Results.objects.create(user_id=int(session_user),
+                               quiz_id=int(current_quiz),
+                               pass_date=datetime.datetime.now(
+                                   timezone.get_current_timezone()),
+                               result=result)
+        del self.request.session[current_quiz]
+
+    def _count_user_answers(self, current_quiz, context):
+        """
+        Method, which count answered questions, initialize context variable
+         which uses for display finish button.
+
+        :param context: 'context' object which is being rendered that view;
+        :var number_of_questions(int): number of questions related with current
+         quiz;
+        :var answers(list): with dict values which contain user answers;
+        :var answered(int): counter, which represent answered answers
+         (not None);
+        :var context['is_can_be_finished']: variable, which uses in template
+         for displaying finish button.
+
+        """
+        number_of_questions = Questions.objects.filter(
+            quiz=current_quiz).count()
+
+        if current_quiz in self.request.session.keys():
+            answers = dict(self.request.session[current_quiz]).values()
```

Comment:
Are you sure you need to use dict method?

Reply:
done.